service/kernel/oneapi/reducer.cpp

```diff
@@ -0,0 +1,205 @@
+                              uint32_t nVectors, uint32_t vectorSize, uint32_t numWorkItems, uint32_t numWorkGroups, Reducer::Result & stepResult,
+                              services::Status * status)
+{
+    auto reduce_kernel = kernelFactory.getKernel("reduce_step_colmajor");
+
+    KernelRange localRange(numWorkItems);
+    KernelRange globalRange(numWorkGroups * numWorkItems);
+
+    KernelNDRange range(1);
+    range.global(globalRange, status);
+    DAAL_CHECK_STATUS_PTR(status);
+    range.local(localRange, status);
+    DAAL_CHECK_STATUS_PTR(status);
+
+    KernelArguments args(4);
+
+    args.set(0, vectors, AccessModeIds::read);
+    args.set(1, nVectors);
+    args.set(2, vectorSize);
+    args.set(3, stepResult.reduceRes, AccessModeIds::write);
+
+    context.run(range, reduce_kernel, args, status);
+}
+
+void Reducer::runFinalStepRowmajor(ExecutionContextIface & context, ClKernelFactoryIface & kernelFactory, Reducer::Result & stepResult,
+                                   uint32_t nVectors, uint32_t vectorSize, uint32_t workItemsPerGroup, Reducer::Result & result,
+                                   services::Status * status)
+{
+    auto reduce_kernel = kernelFactory.getKernel("reduce_final_step_rowmajor");
+
+    KernelRange localRange(workItemsPerGroup);
+    KernelRange globalRange(workItemsPerGroup * nVectors);
+
+    KernelNDRange range(1);
+    range.global(globalRange, status);
+    DAAL_CHECK_STATUS_PTR(status);
+    range.local(localRange, status);
+    DAAL_CHECK_STATUS_PTR(status);
+
+    KernelArguments args(4);
+    args.set(0, stepResult.reduceRes, AccessModeIds::read);
+    args.set(1, nVectors);
+    args.set(2, vectorSize);
+    args.set(3, result.reduceRes, AccessModeIds::write);
+
+    context.run(range, reduce_kernel, args, status);
+}
+
+Reducer::Result Reducer::reduce(const BinaryOp op, Layout vectorsLayout, const UniversalBuffer & vectors, uint32_t nVectors, uint32_t vectorSize,
+                                services::Status * status)
+{
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+    Result result(context, nVectors, vectors.type(), status);
+    return Reducer::reduce(op, vectorsLayout, vectors, result.reduceRes, nVectors, vectorSize, status);
+}
+
+Reducer::Result Reducer::reduce(const BinaryOp op, Layout vectorsLayout, const UniversalBuffer & vectors, UniversalBuffer & resReduce,
+                                uint32_t nVectors, uint32_t vectorSize, services::Status * status)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(Reducer);
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+
+    Result result(context, resReduce, nVectors, vectors.type(), status);
+
+    auto & kernelFactory = context.getClKernelFactory();
+
+    buildProgram(kernelFactory, op, vectors.type());
+
+    const uint32_t maxWorkItemsPerGroup = 256;
```

Comment:
Yes, I agree with you. But we have **all the algorithms** now have constant for device. I think it would be more correct to go one person and fix it in the same style everywhere.